apps/kg/models/pytorch/tensor_models.py

```diff
@@ -76,33 +191,70 @@ def update(self, gpu_id=-1):
                 # the update is non-linear so indices must be unique
                 grad_indices = idx
                 grad_values = grad
-
-                grad_sum = (grad_values * grad_values).mean(1)
-                device = self.state_sum.device
-                if device != grad_indices.device:
-                    grad_indices = grad_indices.to(device)
-                if device != grad_sum.device:
-                    grad_sum = grad_sum.to(device)
-                self.state_sum.index_add_(0, grad_indices, grad_sum)
-                std = self.state_sum[grad_indices]  # _sparse_mask
-                if gpu_id >= 0:
-                    std = std.cuda(gpu_id)
-                std_values = std.sqrt_().add_(1e-10).unsqueeze(1)
-                tmp = (-clr * grad_values / std_values)
-                if tmp.device != device:
-                    tmp = tmp.to(device)
-                # TODO(zhengda) the overhead is here.
-                self.emb.index_add_(0, grad_indices, tmp)
+                if self.async_q is not None:
+                    grad_indices.share_memory_()
+                    grad_values.share_memory_()
+                    self.async_q.put((grad_indices, grad_values, gpu_id))
+                else:
+                    grad_sum = (grad_values * grad_values).mean(1)
+                    device = self.state_sum.device
+                    if device != grad_indices.device:
+                        grad_indices = grad_indices.to(device)
+                    if device != grad_sum.device:
+                        grad_sum = grad_sum.to(device)
+                    self.state_sum.index_add_(0, grad_indices, grad_sum)
+                    std = self.state_sum[grad_indices]  # _sparse_mask
+                    if gpu_id >= 0:
+                        std = std.cuda(gpu_id)
+                    std_values = std.sqrt_().add_(1e-10).unsqueeze(1)
+                    tmp = (-clr * grad_values / std_values)
+                    if tmp.device != device:
+                        tmp = tmp.to(device)
+                    # TODO(zhengda) the overhead is here.
+                    self.emb.index_add_(0, grad_indices, tmp)
```

Comment:
looks like this piece of code is replicated?

Reply:
relplication of which one? you means the async update?
I think it is OK. More thing will be added here if we support soft relation partition. I am working on it.